### PR TITLE
feat: `merodb` migrations

### PR DESCRIPTION
**Pull request by sh3ll3x3c**
# [merodb] Add comprehensive migration pipeline for RocksDB data migration

## Description

This PR implements a complete YAML-driven migration pipeline for `merodb` that enables safe, structured data migration between RocksDB instances. The system supports copying, deleting, and upserting data with rich filtering capabilities, comprehensive verification mechanisms, and automated backup management.

**Key capabilities added:**

- **YAML-based migration plans** with strongly-typed schema (`version: 1`) supporting four step types: `copy`, `delete`, `upsert`, and `verify`
- **Dry-run mode (default)** that previews all operations without making changes, showing key counts, samples, and warnings
- **Apply mode** (`--apply` flag) that executes migrations with atomic WriteBatch operations (1000 keys per batch)
- **Automated backup system** using RocksDB's native backup engine with configurable backup directories via `--backup-dir` flag or plan configuration
- **Pre-flight validation** that runs comprehensive database health checks before executing migrations (foreign key validation, orphaned records detection, etc.)
- **Rich filtering system** supporting context IDs, raw key prefixes, state key prefixes, key ranges, and alias names
- **Verification steps** with multiple assertion types: expected/min/max counts, contains/missing key checks
- **Progress logging** for long-running operations with real-time updates
- **JSON reporting** (`--report <file>`) for machine-readable execution logs in both modes
- **Comprehensive error handling** with validation at plan load time and detailed warnings during execution
- **User-friendly output** displaying effective resolved paths instead of YAML placeholders

**Implementation highlights:**

- Restructured CLI from flags to `clap::Subcommand` architecture (`schema`, `export`, `validate`, `export-dag`, `migrate`)
- Migration context with lazy ABI loading via `OnceCell` for optional WASM file support
- Filter resolution engine that merges defaults with step-level overrides
- Atomic operations using RocksDB `WriteBatch` API for data integrity
- Integrated validation system that prevents migrations on corrupted databases
- Automatic target database directory creation if missing
- Backup module with timestamped incremental backups using RocksDB's BackupEngine
- CLI override system allowing backup control via `--backup-dir` and `--no-backup` flags
- Extensive test infrastructure with `DbFixture` utilities and 30+ test cases

**Files changed:**
- ~8,900 lines added across 27 files
- New modules: `migration/*` (plan, dry_run, execute, context, filters, verification, report, loader, cli, backup)
- Enhanced modules: `validation` (integrated pre-flight checks), `types` (Column parsing)
- Documentation: `MIGRATIONS.md` (implementation plan), `migration/README.md` (detailed reference)

This addresses the need for safe, auditable, and repeatable data migrations in Calimero's RocksDB storage layer with production-ready safeguards.

## Test plan

**Test coverage:**

1. **Unit tests (30+ test cases):**
   - Plan validation (unsupported versions, malformed filters, empty steps)
   - Filter resolution (context IDs, prefixes, ranges, combinations)
   - All step types (copy, delete, upsert, verify)
   - All verification assertions (expected/min/max count, contains/missing key)
   - Edge cases (empty databases, malformed keys, filter mismatches)
   - Multiple column types (State, Meta, Config, Identity, Delta)
   - Execution engine with actual write operations
   - Backup creation and restoration scenarios

2. **Integration tests:**
   - End-to-end dry-run with temporary RocksDB fixtures
   - Execution engine with validation guards
   - JSON report generation for both modes
   - Backup directory creation and incremental backups
   - Target database auto-creation

3. **Manual testing workflow:**
   ```bash
   # Build the project
   cargo build --release

   # Step 1: Preview migration (dry-run mode with resolved paths)
   ./target/release/merodb migrate --plan examples/test_plan.yaml --target-db /tmp/target_db

   # Step 2: Execute migration with automatic backup
   ./target/release/merodb migrate --plan examples/test_plan.yaml --target-db /tmp/target_db --apply

   # Step 3: Execute with custom backup directory
   ./target/release/merodb migrate --plan examples/test_plan.yaml --target-db /tmp/target_db --apply --backup-dir /tmp/backups

   # Step 4: Verify with JSON report
   ./target/release/merodb migrate --plan examples/test_plan.yaml --target-db /tmp/target_db --apply --report results.json

   # Step 5: Skip backup (use with caution)
   ./target/release/merodb migrate --plan examples/test_plan.yaml --target-db /tmp/target_db --apply --no-backup
   ```

**To verify the changes:**
- Run `cargo test` to execute all unit and integration tests
- Check `cargo clippy` for linting (all warnings resolved)
- Run `cargo fmt --check` for formatting compliance
- Verify CLI help text: `cargo run -- migrate --help`
- Test dry-run mode with various YAML plans
- Test apply mode with automatic backup creation
- Verify pre-flight validation catches database corruption
- Test backup directory override functionality

All tests pass, including comprehensive edge case coverage. No UI changes.

## Documentation update

**Documentation added in this PR:**

1. **`tools/merodb/MIGRATIONS.md`** – Implementation plan with step-by-step progress tracking and usage examples
2. **`tools/merodb/src/migration/README.md`** – Detailed reference covering:
   - High-level flow diagrams
   - YAML plan schema and field reference
   - Filter types and step variants
   - Dry-run engine internals
   - Execution engine architecture with backup system
   - Validation integration and pre-flight checks
   - Testing strategy and coverage
   - CLI usage examples and best practices
   - Backup and recovery procedures

3. **`tools/merodb/README.md`** – Updated with migration command overview

**Public documentation updates needed (within 1 day of merge):**

1. Add example migration plans to `tools/merodb/examples/` directory:
   - Basic copy operation with backup
   - Multi-step migration with verification
   - Complex filtering scenarios
   - Backup configuration examples

2. If merodb documentation exists elsewhere in the project, add references to the new migration capabilities including backup management

### CODE REVIEW

**Comment by sh3ll3x3c** [comment]
Screenshot 2025-11-07 at 13 08 46
Screenshot 2025-11-07 at 13 09 03
Screenshot 2025-11-07 at 13 10 04

File: tools/merodb/src/migration/context.rs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+        );
+
+        let db = open_database(&path)?;
+
+        Ok(Self {
+            path,
+            backup_dir,
+            db,
+            read_only: true,
+        })
```
[the remark this answers is not in the document]
Comment (chefsale) [reply tools/merodb/src/migration/context.rs:225]: @sh3ll3x3c

[the remark this answers is not in the document]
Comment (sh3ll3x3c) [reply tools/merodb/src/migration/context.rs:225]: fixing this in the next PR
